diesel/src/pg/types/multirange.rs

```diff
@@ -0,0 +1,111 @@
+        }
+    };
+}
+multirange_has_sql_type!(Datemultirange, 4535, 6155);
+multirange_has_sql_type!(Int4multirange, 4451, 6150);
+multirange_has_sql_type!(Int8multirange, 4536, 6157);
+multirange_has_sql_type!(Nummultirange, 4532, 6151);
+multirange_has_sql_type!(Tsmultirange, 4533, 6152);
+multirange_has_sql_type!(Tstzmultirange, 4534, 6153);
+
+macro_rules! multirange_as_expression {
+    ($ty:ty, $sql_type:ty) => {
+        #[cfg(feature = "postgres_backend")]
+        // this simplifies the macro implementation
+        // as some macro calls use this lifetime
+        #[allow(clippy::extra_unused_lifetimes)]
+        impl<'a, 'b, ST: 'static, T> AsExpression<$sql_type> for $ty {
+            type Expression = SqlBound<$sql_type, Self>;
+            fn as_expression(self) -> Self::Expression {
+                SqlBound::new(self)
+            }
+        }
+    };
+}
+
+multirange_as_expression!(&'a [(Bound<T>, Bound<T>)], Multirange<ST>);
+multirange_as_expression!(&'a [(Bound<T>, Bound<T>)], Nullable<Multirange<ST>>);
+multirange_as_expression!(&'a &'b [(Bound<T>, Bound<T>)], Multirange<ST>);
+multirange_as_expression!(&'a &'b [(Bound<T>, Bound<T>)], Nullable<Multirange<ST>>);
+multirange_as_expression!(Vec<(Bound<T>, Bound<T>)>, Multirange<ST>);
+multirange_as_expression!(Vec<(Bound<T>, Bound<T>)>, Nullable<Multirange<ST>>);
+multirange_as_expression!(&'a Vec<(Bound<T>, Bound<T>)>, Multirange<ST>);
+multirange_as_expression!(&'a Vec<(Bound<T>, Bound<T>)>, Nullable<Multirange<ST>>);
+multirange_as_expression!(&'a &'b Vec<(Bound<T>, Bound<T>)>, Multirange<ST>);
+multirange_as_expression!(&'a &'b Vec<(Bound<T>, Bound<T>)>, Nullable<Multirange<ST>>);
+
+#[cfg(feature = "postgres_backend")]
+impl<T, ST> FromSql<Multirange<ST>, Pg> for Vec<(Bound<T>, Bound<T>)>
+where
+    T: FromSql<ST, Pg>,
+{
+    fn from_sql(value: PgValue<'_>) -> deserialize::Result<Self> {
+        let mut bytes = value.as_bytes();
+        let len = bytes.read_i32::<NetworkEndian>()?;
+
+        (0..len)
+            .map(|_| {
+                let range_size = bytes.read_i32::<NetworkEndian>()?;
+                let (range_bytes, new_bytes) = bytes.split_at(range_size as usize);
+                bytes = new_bytes;
+                FromSql::from_sql(PgValue::new_internal(range_bytes, &value))
+            })
+            .collect()
+    }
+}
+
+#[cfg(feature = "postgres_backend")]
+impl<T, ST> ToSql<Multirange<ST>, Pg> for [(Bound<T>, Bound<T>)]
+where
+    T: ToSql<ST, Pg>,
+{
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> serialize::Result {
+        out.write_u32::<NetworkEndian>(self.len() as u32)?;
+
+        let mut buffer = Vec::new();
+        for value in self {
+            {
+                let mut inner_buffer = Output::new(ByteWrapper(&mut buffer), out.metadata_lookup());
+                ToSql::<Range<ST>, Pg>::to_sql(&value, &mut inner_buffer)?;
+            }
+            out.write_u32::<NetworkEndian>(buffer.len() as u32)?;
```

Comment:
It might be worth to convert that into a failable cast via `TryInto`, just to make sure that we don't overflow the u32 here. It might be an edge case, but better be save than sorry. (I really should check the other variable sized `ToSql` impls for such issues, at some point)

Reply:
I agree, I copied from the Range implementation it should have been implemented before `TryFrom` was stable.

Reply:
[I fix this in range too](https://github.com/diesel-rs/diesel/pull/4159/commits/f75e5d856eef3d8b60e2d2f366215a22ea78816d)****

Reply:
Thanks. The array `ToSql` impl does seem to suffer from the same problem: https://github.com/diesel-rs/diesel/blob/c4822088da5b1cb51164cec2bfe3ca0864e32e70/diesel/src/pg/types/array.rs#L104

Reply:
Thanks, I open a PR for that: https://github.com/diesel-rs/diesel/pull/4166/files